apps/chat/models.py

```diff
@@ -104,3 +113,107 @@ def _get_langchain_dict(self, content, message_type):
+    trigger_event = models.CharField(choices=TriggerEvent.choices, db_index=True, blank=False)
+    recurring = models.BooleanField()
+    time_period = models.CharField(choices=TimePeriod.choices)
+    frequency = models.IntegerField(default=1)
+    repetitions = models.IntegerField(default=0)
+    prompt_text = models.TextField()
+
+    def save(self, *args, **kwargs):
+        if self.recurring and self.repetitions == 0:
+            raise ValueError(_("Recurring schedules require `repetitions` to be larger than 0"))
+        if not self.recurring and self.repetitions > 0:
+            raise ValueError(_("Non recurring schedules cannot have `repetitions` larger than 0"))
+        if self.id:
+            self.update_scheduled_messages()
+        return super().save(*args, **kwargs)
+
+    def update_scheduled_messages(self):
+        """
+        This method updates the scheduled_messages queryset by considering the following criteria:
+        - trigger_event and recurring: No effect. Users will not be able to change these at the moment
+
+        - Number of repetitions:
+            - If new repetitions are greater than total_triggers, mark resolved ones as unresolved.
+            - If new repetitions are less than total_triggers, mark unresolved ones as resolved.
+
+        - Frequency and time period (delta change):
+            - If the scheduled message's last_triggered_at field is None (it has not fired), the created_at field
+            is used as the baseline for adding the new delta
+            - If the scheduled message's last_triggered_at field is not None (it has fired before), that field is
+            then used as the baseline for adding the new delta
+        """
+        (
+            self.scheduled_messages.annotate(
+                new_delta=MakeInterval(self.time_period, self.frequency),
+            ).update(
+                resolved=Case(
+                    When(total_triggers__lt=self.repetitions, then=False),
+                    When(total_triggers__gte=self.repetitions, then=True),
+                    output_field=models.BooleanField(),
+                ),
+                next_trigger_date=Case(
+                    When(last_triggered_at__isnull=True, then=F("created_at") + F("new_delta")),
+                    When(last_triggered_at__isnull=False, then=F("last_triggered_at") + F("new_delta")),
+                    output_field=DateTimeField(),
+                ),
+            )
+        )
+
+
+class ScheduledMessage(BaseTeamModel):
+    schedule = models.ForeignKey(ScheduledMessageConfig, on_delete=models.CASCADE, related_name="scheduled_messages")
+    participant = models.ForeignKey(
+        "experiments.Participant", on_delete=models.CASCADE, related_name="schduled_messages"
+    )
+    next_trigger_date = models.DateTimeField(null=True)
+    last_triggered_at = models.DateTimeField(null=True)
+    total_triggers = models.IntegerField(default=0)
+    resolved = models.BooleanField(default=False)
```

Comment:
wondering about renaming this to 'complete'

Reply:
Hmm..agreed, but maybe `is_complete` is a bit clearer than `complete`

Reply:
https://github.com/dimagi/open-chat-studio/pull/373/commits/6996e196dc707b432776eb8dc8b6312948cd14a4